torsiondrive/dihedral_scanner.py

```diff
@@ -166,13 +169,79 @@ def build_dihedral_mask(self, dihedral_ranges):
     #--------------------
     #  General methods
     #--------------------
+
+    def measure_dihedrals(self, molecule, dihedral_list, check_linear = True, check_bonded = True):
+        """
+        Requested features: check angle  of each dihedral, raise a warning message if the angle is larger than 165.
+        Aim to replace molecule.measure_dihedrals in "get_dihedral_id."
+        """
+        dihedrals = []
+        for lst in dihedral_list:
+            i, j, k, l = lst
+
+            if check_bonded == True:
+                bond_list = []
+                if i > j:
+                    bond_list.append((j, i))
+                else:
+                    bond_list.append((i, j))
+                if j > k:
+                    bond_list.append((k, j))
+                else:
+                    bond_list.append((j, k))
+                if k > l:
+                    bond_list.append((l, k))
+                else:
+                    bond_list.append((k, l))
+
+                if 'bonds'  not in molecule.Data:
+                    molecule.build_bonds()
+                if any(p not in molecule.bonds for p in bond_list):
+                    print([p not in molecule.bonds for p in bond_list])
+                    warn(UserWarning(f"Warning! some bonds in the dihedral {lst} is missing!"))
+
+            x1 = molecule.xyzs[0][i]
+            x2 = molecule.xyzs[0][j]
+            x3 = molecule.xyzs[0][k]
+            x4 = molecule.xyzs[0][l]
+            v1 = x2-x1
+            v2 = x3-x2
+            v3 = x4-x3
+
+            if check_linear == True:
+                cos_angle1 = np.dot(-v1, v2)/(np.linalg.norm(v1)*np.linalg.norm(v2))
+                cos_angle2 = np.dot(-v2, v3)/(np.linalg.norm(v2)*np.linalg.norm(v3))
+                angle1 = np.arccos(cos_angle1) * 180 / np.pi
+                angle2 = np.arccos(cos_angle2) * 180 / np.pi
+                if angle1 > 165 or angle2 > 165:
+                    warn(UserWarning(f"Warning! {lst} contains a straight angle!"))
+
+
+            n1 = np.cross(v1, v2)
+            norm1 = np.linalg.norm(n1)
+            if norm1 != 0:
+                n1 /= norm1
+            n2 = np.cross(v2, v3)
+            norm2 = np.linalg.norm(n2)
+            if norm2 != 0:
+                n2 /= norm2
+            m1 = np.cross(v2, n1) / np.linalg.norm(v2)
```

Comment:
Can you make sure this definition is equivalent to the QCElemental/geometric definition? I know Lee-Ping and I were opposite at one point.

Reply:
We are planning to reproduce the values of the `Molecule.measure_dihedrals()` function so it will keep the result consistent.